### PR TITLE
Issue-91 [FEATURE] Removes preloading of blocks on edit post screen

### PR DESCRIPTION
**Pull request by mslinnea**
## Summary

WordPress pre-loads all Reusable Blocks (Synced Patterns) on the edit post screen, site editor, and widget screen. For sites with hundreds of blocks, this is a lot of processing overhead, as the_content and all other filters run on these blocks. Disabling this behavior make the editing experience faster and less fragile for sites with a large number of blocks.

## Notes for reviewers

None.

## Other Information

- [x] I updated the `README.md` file for any new/updated features.
- [x] I updated the `CHANGELOG.md` file for any new/updated features.

### CODE REVIEW

File: src/alley/wp/alleyvate/features/class-disable-block-editor-rest-api-preload-paths.php [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+	public static function filter__block_editor_rest_api_preload_paths( $paths ) {
+		if ( ! is_array( $paths ) ) {
+			return [];
+		}
+		return array_values(
+			array_filter(
+				$paths,
+				function ( $v ) {
+					// Remove the blocks preload path for performance reasons.
+					return ! is_string( $v ) || ! str_starts_with( $v, '/wp/v2/blocks?context=edit' );
```
Comment (dlh01) [inline src/alley/wp/alleyvate/features/class-disable-block-editor-rest-api-preload-paths.php:50]: Is it best to wipe out preloading synced patterns on all sites? Is it ever beneficial for sites with fewer synced patterns?

Comment (mslinnea) [reply src/alley/wp/alleyvate/features/class-disable-block-editor-rest-api-preload-paths.php:50]: I wish I could find a discussion about the possible benefit as I certainly don't want to cause any issues with this change.  I'll ask around.

I do know that if only some blocks are pre-loaded (like only 10) that there are issues with searching for reusable blocks.  So it seems that it has to be either an all or nothing approach.

Comment (dlh01) [reply src/alley/wp/alleyvate/features/class-disable-block-editor-rest-api-preload-paths.php:50]: Yeah, the all-or-nothing part doesn't seem like a big deal to me, rather, should this logic kick in only after a site gets more than 20 (or whatever) synced patterns? And that number could be filterable, etc.

Comment (kevinfodness) [reply src/alley/wp/alleyvate/features/class-disable-block-editor-rest-api-preload-paths.php:50]: imo it's a sensible default to turn it off completely as long as it doesn't cause any issues with synced patterns/reusable blocks, and if individual sites want to preload, they can turn off the feature with the filter as @mogmarsh said.

Comment (mslinnea) [reply src/alley/wp/alleyvate/features/class-disable-block-editor-rest-api-preload-paths.php]: In addition to testing this on a very large traditional site, I also tested it on a block theme and had no issue using synced patterns.

File: src/alley/wp/alleyvate/features/class-disable-block-editor-rest-api-preload-paths.php [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+	public static function filter__block_editor_rest_api_preload_paths( $paths ) {
+		if ( ! \is_array( $paths ) ) {
+			return [];
+		}
+		return array_values(
+			array_filter(
+				$paths,
+				function ( $v ) {
+					// Remove the blocks preload path for performance reasons.
+					return ! \is_string( $v ) || ! str_starts_with( $v, '/wp/v2/blocks?context=edit' );
```
Comment (dlh01) [inline src/alley/wp/alleyvate/features/class-disable-block-editor-rest-api-preload-paths.php:50]: Checking that `$v` is a string before passing it to `str_starts_with()` makes sense, but removing any non-string value from `$paths` seems to me to go beyond the stated scope of this feature. Is that behavior intentional?

File: src/alley/wp/alleyvate/features/class-disable-block-editor-rest-api-preload-paths.php [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+	/**
+	 * Filter the block editor REST API preload paths.
+	 *
+	 * @param mixed[] $paths The paths to preload.
+	 *
+	 * @return mixed[] The filtered paths.
+	 */
+	public static function filter__block_editor_rest_api_preload_paths( $paths ) {
+		if ( ! \is_array( $paths ) ) {
+			return [];
```
Comment (dlh01) [inline src/alley/wp/alleyvate/features/class-disable-block-editor-rest-api-preload-paths.php:43]: Sort of in line with the other comment, is it intentional to overwrite `$paths` when it's a non-array?

File: tests/alley/wp/alleyvate/features/test-disable-block-editor-rest-api-preload-paths.php [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+			sprintf( '/wp/v2/types/%s?context=edit', $post_type ),
+			'/wp/v2/users/me',
+			[ rest_get_route_for_post_type_items( 'attachment' ), 'OPTIONS' ],
+			[ rest_get_route_for_post_type_items( 'page' ), 'OPTIONS' ],
+			[ rest_get_route_for_post_type_items( 'wp_block' ), 'OPTIONS' ],
+			[ rest_get_route_for_post_type_items( 'wp_template' ), 'OPTIONS' ],
+			sprintf( '%s/autosaves?context=edit', $rest_path ),
+			'/wp/v2/settings',
+			[ '/wp/v2/settings', 'OPTIONS' ],
+		];
```
Comment (dlh01) [inline tests/alley/wp/alleyvate/features/test-disable-block-editor-rest-api-preload-paths.php:76]: All of these paths are being added to mimic what usually happens in `wp-admin`, but it's a little hard to follow what they're meant to demonstrate.

If the intention is to ensure that `/wp/v2/blocks?context=edit&per_page=-1` is removed from the list of paths, it seems to me that that could be shown without having to declare all these other paths.

If the intention is to ensure that the other paths are _not_ removed from the list of paths, that behavior doesn't seem to be tested.

File: tests/alley/wp/alleyvate/features/test-disable-block-editor-rest-api-preload-paths.php [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+		$this->assertContains( '/wp/v2/blocks?context=edit&per_page=-1', $preload_paths );
+
+		// Activate feature.
+		$this->feature->boot();
+
+		// Apply the filter that modifies the preload paths.
+		$preload_paths = apply_filters( 'block_editor_rest_api_preload_paths', $preload_paths ); // phpcs:ignore WordPress.NamingConventions.PrefixAllGlobals.NonPrefixedHooknameFound
+
+		// The blocks rest path should no longer be in the preload paths.
+		$this->assertNotContains( '/wp/v2/blocks?context=edit&per_page=-1', $preload_paths );
```
Comment (dlh01) [inline tests/alley/wp/alleyvate/features/test-disable-block-editor-rest-api-preload-paths.php:88]: I think it would be dispositive to apply the filter before booting the feature and verify that the blocks REST path is still in the list. That would go towards demonstrating that it was the Alleyvate feature that removed the path from the list, not some other code that was running already.

**Review by dlh01 — APPROVED** [review]
🌳 Some thoughts about how this feature could be tightened up, but no blocking feedback.

File: tests/alley/wp/alleyvate/features/test-disable-block-editor-rest-api-preload-paths.php (hunk 5 above)
Comment (mslinnea) [reply tests/alley/wp/alleyvate/features/test-disable-block-editor-rest-api-preload-paths.php:94]: Good point!  I added this in https://github.com/alleyinteractive/wp-alleyvate/pull/92/commits/e89e2435f4fa4c55f595f244daa3adc1ab0c5629

File: tests/alley/wp/alleyvate/features/test-disable-block-editor-rest-api-preload-paths.php (hunk 4 above)
Comment (mslinnea) [reply tests/alley/wp/alleyvate/features/test-disable-block-editor-rest-api-preload-paths.php:76]: I added assertions to verify that the other paths are _not_ removed.

File: src/alley/wp/alleyvate/features/class-disable-block-editor-rest-api-preload-paths.php (hunk 3 above)
Comment (mslinnea) [reply src/alley/wp/alleyvate/features/class-disable-block-editor-rest-api-preload-paths.php]: Excellent catch! I've updated this to just return $paths in that case.

File: src/alley/wp/alleyvate/features/class-disable-block-editor-rest-api-preload-paths.php (hunk 2 above)
Comment (mslinnea) [reply src/alley/wp/alleyvate/features/class-disable-block-editor-rest-api-preload-paths.php:50]: This only removes strings that start with `/wp/v2/blocks?context=edit`.  I've added tests to verify this behavior.

Comment (dlh01) [reply src/alley/wp/alleyvate/features/class-disable-block-editor-rest-api-preload-paths.php:50]: Ah, I misread! All good.

**Review by dlh01 — APPROVED** [review]
🌳
